tests/MongoDB.Driver.Core.Tests/Core/Authentication/AwsSignatureVersion4Tests.cs

```diff
@@ -0,0 +1,181 @@
+                salt,
+                host,
+                out var actualAuthorizationHeader,
+                out var actualTimestamp);
+
+            actualAuthorizationHeader.Should().Be(expectedAuthorizationHeader);
+            actualTimestamp.Should().Be(expectedTimestamp);
+        }
+
+        [Fact]
+        public void GetCanonicalHeaders_should_return_expected_result()
+        {
+            var timestamp = new DateTime(2020, 03, 12, 14, 23, 46).ToString("yyyyMMddTHHmmssZ");
+            var requestHeaders = new SortedDictionary<string, string>
+            {
+                ["X-MongoDB-GS2-CB-Flag"] = "n",
+                ["Content-Type"] = "application/x-www-form-urlencoded",
+                ["X-Amz-Date"] = timestamp,
+                ["Content-Length"] = "42",
+                ["Host"] = "iam.testhost.com",
+                ["X-MongoDB-Server-Nonce"] = "123",
+                ["X-Amz-Security-Token"] = "321"
+            };
+            var expected = "content-length:42\n" +
+                "content-type:application/x-www-form-urlencoded\n" +
+                "host:iam.testhost.com\n" +
+                "x-amz-date:20200312T142346Z\n" +
+                "x-amz-security-token:321\n" +
+                "x-mongodb-gs2-cb-flag:n\n" +
+                "x-mongodb-server-nonce:123";
+
+            var actual = AwsSignatureVersion4Reflector.GetCanonicalHeaders(requestHeaders);
+
+            actual.Should().Be(expected);
+        }
+
+        [Theory]
+        [InlineData("sts.amazonaws.com", "us-east-1")]
+        [InlineData("first", "us-east-1")]
+        [InlineData("first.second", "second")]
+        [InlineData("first.second.third", "second")]
+        public void GetRegion_should_return_expected_result(string host, string expectedRegion)
+        {
+            var region = AwsSignatureVersion4Reflector.GetRegion(host);
+
+            region.Should().Be(expectedRegion);
+        }
+
+        [Fact]
+        public void GetSignedHeaders_should_return_expected_result()
+        {
+            var timestamp = new DateTime(2020, 03, 12, 14, 23, 46).ToString("yyyyMMddTHHmmssZ");
+            var requestHeaders = new SortedDictionary<string, string>
+            {
+                ["X-Amz-Security-Token"] = "321",
+                ["Content-Type"] = "application/x-www-form-urlencoded",
+                ["X-Amz-Date"] = timestamp,
+                ["Content-Length"] = "42",
+                ["Host"] = "iam.testhost.com",
+                ["X-MongoDB-GS2-CB-Flag"] = "n",
+                ["X-MongoDB-Server-Nonce"] = "123"
+            };
+            var expected = "content-length;content-type;host;x-amz-date;x-amz-security-token;x-mongodb-gs2-cb-flag;x-mongodb-server-nonce";
+
+            var actual = AwsSignatureVersion4Reflector.GetSignedHeaders(requestHeaders);
+
+            actual.Should().Be(expected);
+        }
+
+        // private methods
+        private static SecureString ToSecureString(string str)
```

Comment:
Yet another copy of `ToSecureString`?

Maybe we need a new helper class:

    public static class SecureStringHelper
    {
        public static string ToInsecureString(SecureString value);
        public static SecureString ToSecureString(string value);
    }

Note: normally it is better to use the `DecryptedSecureString` class rather than calling a method like `ToInsecureString`.

Ideally this class would be public in Core (or Bson?) and accessible to all projects, but we often have discussions about whether something can be public or not. We might have to make it `internal` and put this in `MongoDB.Shared` and link to it from each project that wants to use it.

Reply:
Done.